ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml

```diff
@@ -359,26 +359,39 @@ let dss_loadavg () =
     )
   ]
 
+let rec sum acc n f =
+  match n with n when n >= 0 -> sum (acc + f n) (n - 1) f | _ -> acc
+
+let count_vcpus pred domains xc =
```

Comment:
In light of the discussion of `count_running_domain`, could this be done in a simpler, similar way? The predicate `ctrl.Xenctrl.online && not ctrl.Xenctrl.blocked` could be implemented in this function and not passed in from the outside and this would result in one compact function.

Reply:
Some other statuses could be added in the future, such as `ctrl.Xenctrl.blocked `or `ctrl.Xenctrl.running`. Therefore, I believe keeping the `count_running_domain` will be beneficial for compatibility

Reply:
The `xc` can be wrapped into `pred` (renamed to `count_vcpus_in_domain`).
So basically the functions could be:
```
let count_vcpus_in_domain domain =
  let dom, _, domid = domain in
  let f vcpu_id =
    let open Xenctrl in
    let vcpuinfo = domain_get_vcpuinfo xc domid vcpu_id in
    if vcpuinfo.online && not vcpuinfo.blocked then 1 else 0
  in
  sum 0 dom.Xenctrl.max_vcpu_id f
in
let vcpus =
  List.map count_vcpus_in_domain domains
  |> List.fold_left (+) 0
in

let count_running_domain domain =
  let dom, _, _ = domain in
  if dom.Xenctrl.running then 1 else 0
in
let running_domains =
  List.map count_running_domain domains
  |> List.fold_left (+) 0
```